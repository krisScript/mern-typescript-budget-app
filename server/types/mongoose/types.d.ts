namespace Mongoose {
  interface Query<T> {
    cache(options: CacheOptions): Query<T>;
    useCache: boolean;
    hashKey: string;
  }
}
type CacheOptions = {
  key?: string;
};
