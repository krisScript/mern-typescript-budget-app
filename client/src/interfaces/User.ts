interface User {
  username: string;
  email: string;
  userId: string;
}
export default User;
