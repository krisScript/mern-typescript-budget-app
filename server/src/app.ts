import express from 'express';
import { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import mongoose, { Error } from 'mongoose';
import authRoutes from './routes/auth';
import expenseRoutes from './routes/expense';
import helmet from 'helmet';
import rateLimiterRedisMiddleware from './middleware/rateLimiterRedisMiddleware';
import cache from './services/cache';
const app = express();

app.use(helmet());
// app.use(rateLimiterRedisMiddleware);
app.use(bodyParser.json()),
  app.use(
    (req: Request, res: Response, next: NextFunction): void => {
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader(
        'Access-Control-Allow-Methods',
        'OPTIONS, GET, POST, PUT, PATCH, DELETE',
      );
      res.setHeader(
        'Access-Control-Allow-Headers',
        'Content-Type, Authorization',
      );
      next();
    },
  ),
  app.use('/auth', authRoutes);
app.use(expenseRoutes);

app.use(
  (error: any, req: Request, res: Response, next: NextFunction): void => {
    const status = error.status || 500;
    console.log(error);
    if (error.data) {
      const { data } = error;
      res.status(status).json({ data });
    } else {
      const { msg } = error;
      res.status(status).json({ msg });
    }
  },
);
export default app;
