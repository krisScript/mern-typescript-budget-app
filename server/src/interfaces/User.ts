import { Document } from 'mongoose';
interface User extends Document {
  email: string;
  password: string;
  username: string;
  confirmed: boolean;
}
export default User;
