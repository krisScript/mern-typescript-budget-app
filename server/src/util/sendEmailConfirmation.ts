import jwt from 'jsonwebtoken';
import sendEmail from './sendEmail';
import MailOptions from '../interfaces/MailOptions';
const sendEmailConfirmation = (userId: string, email: string): void => {
  const emailSecret: any = process.env.EMAIL_SECRET;
  const token = jwt.sign(
    {
      userId,
    },
    emailSecret,
    { expiresIn: '1h' },
  );
  const appEmail: any = process.env.EMAIL;
  const url = `http://localhost:3000/auth/confirmation/${token}`;
  const mailOptions: MailOptions = {
    from: appEmail,
    to: email,
    subject: 'Email confirmation.',
    html: `Please click this email to confirm your email: <a href="${url}">${url}</a>`,
  };
  sendEmail(mailOptions);
};
export default sendEmailConfirmation;
