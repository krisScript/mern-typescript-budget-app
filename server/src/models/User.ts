import mongoose, { Document, Schema, Model, model } from 'mongoose';
import User from '../interfaces/User';

const UserSchema: Schema = new Schema({
  email: { type: String, required: true, unique: true },
  password: { type: String, required: true },
  username: { type: String, required: true },
  confirmed: { type: Boolean, default: false },
});
export default mongoose.model<User>('User', UserSchema);
