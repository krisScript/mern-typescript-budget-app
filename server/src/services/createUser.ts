import User from '../models/User';
import bcrypt from 'bcryptjs';
interface CreateUserResponse {
  userId: string;
}
const createUser = async (
  email: string,
  username: string,
  password: string,
): Promise<CreateUserResponse> => {
  const hashedPassword = await bcrypt.hash(password, 12);
  const user = new User({
    email,
    password: hashedPassword,
    username,
  });
  const userId = user._id.toString();
  await user.save();
  return { userId };
};
export default createUser;
