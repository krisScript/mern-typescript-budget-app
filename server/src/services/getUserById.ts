import User from '../models/User';
import ValidationError from '../classes/ValidationError';
import CustomError from '../classes/CustomError';
import ValidationErrorType from '../interfaces/ValidationErrorType';
import UserType from '../interfaces/User';
const getUserById = async (userId: string): Promise<UserType> => {
  try {
    const user = await User.findById(userId);
    if (!user) {
      const error = new CustomError('User not found!', 404);
      throw error;
    }
    return user;
  } catch (err) {
    const validationErrorObj: ValidationErrorType = {
      location: 'body',
      param: 'email',
      msg: 'User with this email does not exist!',
      value: '',
    };
    const validationError = new ValidationError('Validation Error', 403, [
      validationErrorObj,
    ]);
    throw validationError;
  }
};
export default getUserById;
