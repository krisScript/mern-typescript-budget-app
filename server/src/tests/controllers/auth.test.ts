import request from 'supertest';
import app from '../../app';
import mongoose from 'mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import User from '../../models/User';
import sendEmailConfirmation from '../../services/sendEmailConfirmation';
import createUser from '../../services/createUser';
import getUserById from '../../services/getUserById';
let mongoServer: any;
jasmine.DEFAULT_TIMEOUT_INTERVAL = 600000;
jest.mock('../../services/sendEmailConfirmation');
// jest.mock('../../services/getUserById');

describe('/auth', (): void => {
  let mongoServer: any;
  const opts = {};

  beforeAll(async () => {
    mongoServer = new MongoMemoryServer();
    const mongoURI = await mongoServer.getConnectionString();
    await mongoose.connect(mongoURI, { useNewUrlParser: true });
  });

  afterAll(async () => {
    await mongoServer.stop();
    await mongoose.disconnect();
  });

  describe('/sign-up', (): void => {
    const username = 'test2UserName';
    const email = 'testMail@mail.com';
    const password = '1234567891011';
    const incorrectPassword = 'aaaaaaaaaaaaaa';
    it('signing up new user', async (): Promise<void> => {
      const response = await request(app)
        .post('/auth/sign-up')
        .send({
          email,
          username,
          password,
          matchPassword: password,
        });
      expect(response.status).toEqual(200);
      expect(sendEmailConfirmation).toBeCalledTimes(1);
      expect(response.body).toMatchSnapshot();
    });
    it('signing up new user with used email and username', async (): Promise<
      void
    > => {
      const response = await request(app)
        .post('/auth/sign-up')
        .send({
          email,
          username,
          password,
          matchPassword: password,
        });
      expect(response.error).toMatchSnapshot();
      expect(response.status).toEqual(403);
      expect(response.body).toMatchSnapshot();
    });
    it('signing up new user non matching password', async (): Promise<void> => {
      const response = await request(app)
        .post('/auth/sign-up')
        .send({
          email,
          username,
          password,
          matchPassword: incorrectPassword,
        });
      expect(response.error).toMatchSnapshot();
      expect(response.status).toEqual(403);
      expect(response.body).toMatchSnapshot();
    });
  });
  describe('/login', (): void => {});
});
