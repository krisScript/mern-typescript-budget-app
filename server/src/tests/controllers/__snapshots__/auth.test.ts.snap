// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/auth /sign-up signing up new user 1`] = `
Object {
  "message": "User created!",
}
`;

exports[`/auth /sign-up signing up new user non matching password 1`] = `[Error: cannot POST /auth/sign-up (403)]`;

exports[`/auth /sign-up signing up new user non matching password 2`] = `
Object {
  "data": Array [
    Object {
      "location": "body",
      "msg": " Username is already taken. Please try again.",
      "param": "username",
      "value": "test2UserName",
    },
    Object {
      "location": "body",
      "msg": "Email is already used,please user another one",
      "param": "email",
      "value": "testMail@mail.com",
    },
    Object {
      "location": "body",
      "msg": "Your passwords do not match.",
      "param": "matchPassword",
      "value": "aaaaaaaaaaaaaa",
    },
  ],
}
`;

exports[`/auth /sign-up signing up new user with used email and username 1`] = `[Error: cannot POST /auth/sign-up (403)]`;

exports[`/auth /sign-up signing up new user with used email and username 2`] = `
Object {
  "data": Array [
    Object {
      "location": "body",
      "msg": " Username is already taken. Please try again.",
      "param": "username",
      "value": "test2UserName",
    },
    Object {
      "location": "body",
      "msg": "Email is already used,please user another one",
      "param": "email",
      "value": "testMail@mail.com",
    },
  ],
}
`;
